Core/include/Acts/Utilities/detail/Subspace.hpp

```diff
@@ -88,24 +87,34 @@ class FixedSizeSubspace {
   std::array<uint8_t, kSize> m_axes;
 
  public:
-  /// Construct from an array of axis indices.
+  /// Construct from a container of axis indices.
   ///
-  /// @tparam index_container_t SequenceContainer for indices
+  /// @tparam index_container_t Forward-iterable Container for indices
   /// @param indices Ordered, unique axis indices, must contain kSize entries
   template <typename index_container_t>
   constexpr FixedSizeSubspace(const index_container_t& indices) {
-    // verify inputs
-    for (auto i = 0u; i < kSize; ++i) {
-      assert((indices[i] < kFullSize) and
+    // this way of iteration imposes least constraints on the input container
+    auto idx = std::begin(indices);
+    auto end = std::end(indices);
+    size_t istore = 0u;
+    // limit iteration both by input and storage size
+    for (; (idx != end) and (istore < kSize); ++idx, ++istore) {
+      // verify inputs
+      assert((*idx < kFullSize) and
              "Axis indices must be within the full space");
-      if ((i + 1u) < kSize) {
-        assert((indices[i] < indices[i + 1u]) and
+      if (0u < istore) {
+        assert((m_axes[istore - 1u] < *idx) and
                "Axis indices must be strongly ordered and unique");
       }
+      // assign stored axis
+      m_axes[istore] = static_cast<uint8_t>(*idx);
     }
-    // assign stored axis
-    for (auto i = 0u; i < kSize; ++i) {
-      m_axes[i] = static_cast<uint8_t>(indices[i]);
+    assert((istore == kSize) and "Not enough input indices");
+    assert((idx == end) and "Too many input indices");
+    // if there are missing input indices, fill the axes with invalid values
+    // this is a defensive measure for non-debug builds to provoke segfaults
+    for (; istore < kSize; ++istore) {
+      m_axes[istore] = UINT8_MAX;
```

Comment:
Can we not fail on this? The checks are not that expensive that we need to restrict them to debug. I'm just saying this because this sounds like a nightmare debug session waiting to happen.

Reply:
I would like the `Subspace` class to be `constexpr` and then I can not throw in the constructor. This is a result of the switch to a templated input container discussed above.

Reply:
Hm, I understand. I think this failure mode is insidious enough that I would just pop in an `abort()` here and fail anyway. I don't expect this is something you want to be able to `catch`.

Reply:
I was thinking of another option: switch back to `std::array` as input for the `Subspace` and move the error handling to the `Measurement` where we can throw an exception. That way the `Subspace` helper can not have that error mode.

I am not a big fan of `abort()` in a library. If somebody else calls us, it is not our job to crash their program.

Reply:
@paulgessinger I switched back to requiring `std::array` as input for both `Subspace` and `Measurement`. That removes the potential error case and moves the error checking to the user. In #557 `Measurement` is usually created through `makeMeasurement` anyways, so my initial concern about usability is not quite as strong as I suspected.